### ISSUE TITLE
Include NF-e modules in the localization

### ISSUE THREAD

**Issue by danimaribeiro** [OCA/l10n-brazil#155]
I suggest to include this project in localization:
https://github.com/odoo-brazil/odoo-brazil-eletronic-documents

As this module is essential would be better to be included together

Tests need to be done in those modules, as today it doesn't have any.

**Comment by rvalyi** [comment]
:-1: 

Hello @danimaribeiro sorry but I'm against that for now and I explained why here:
https://github.com/OCA/l10n-brazil/pull/129#issuecomment-110458678

The OCA is about having a high quality codebase we kind of pass the image it's somewhat guaranteed. Sorry but the https://github.com/aricaldeira/PySPED library is NOTHING GRANTED. It's maintained by a single guy (who has all its merit, that's not the question), with very low reactivity (how many pending merge since years), the code quality is really so-so with no test... If you look the 340 files changeset for the 3.1 Nfe serie, you can think what would happen if he stopped maintaining. Honnestly NOBODY here would have done that work. Depending on the pysped would kind of taint the project with a very uncertain dependency.

So I would say before extending what is in the scope of the OCA in this project, we should first ensure that the project contributors extend their work to the OCA instead. Today only me and sometimes Renato do OCA level stuff here. You and @mileo are contributing in this project but that's about it. We first need more people doing real work. The OCA is not about trying to gather all Odoo modules in the world, it's just about having people agree on common parts, the more we expand, the less we are efficient in that common part. Next why not the Magento connector localization in this project? why not PAF-EC, I mean why group everything in a repo?

I really think it's important that each fiscal transmission technology keeps as an independent project, it's a clear boundary separation. Also, each transmission technology can have a different leadership, so for instance the leadership if you and @mileo on the pysped transmission is not questioned at all. But We think it's preferable projects have clear responsible: like us for this one and you for the transmission. It's really common that Github repo are under the responsability of the guy who wrote the code, right? Why try mix everything and think it will be better? Odoo itself is already too much of a monolith, let's don't replicate this error please...

Finally, the pysped transmission project absolutely deserves tests. But I think there is no reason we cannot make them happen. So for instance the ooor Ruby connector I built IS NOT part of the OCA but is using the same Travis settings for tests: 
https://github.com/akretion/ooor/blob/master/.travis.yml

Why not just do the same for the pysped transmission project?

**Comment by danimaribeiro** [comment]
We are trying to maintain this here:
https://github.com/odoo-brazil/PySPED/commits/8.0

Next step would be to create a separate pip package for use in instalations and everything, like any other module odoo depends on.

I still think the NF-e should be in the core, it's essential here in Brazil. There is no need to have different fiscal transmission since the government uses just one standard that's common for all. Maybe the odoo is too much monolith, but some things are just necessary.

I don't think we need to put this right now, it's just an idea, but with the coming of NFC-e, MD-e, this is becoming more essential every day.

**Comment by rvalyi** [comment]
@danimaribeiro it's not because it's not the same project that i's not required. I'm not telling that.
Note that we reference it in the README of this project as we would reference any mature or credible other transmissor (some other projects were attempted.) Look, it's just like say Rails and the RDMS driver, nearly always one is needed but still the aren't bundled into Rails to draw a clear project interface and open the door to several concurrent implementations. We want the same with the transmissor.

I maintain my opposition to get it mixed. the Pysped has other issues. May be some days Odoo migrate to Python 3, or even some folks are starting to run Odoo on Pypy and that would never work with Pysped any soon. We absolutely use it, we absolutely thank you for working in this other project, but we don't want it in the same project. Here the project boundaries are quite clear: just enough to extend the standard Odoo Addons. Need more -> other project, sounds fair no? Well we eventually just make an exception with very central OCA modules we extracted from the localization.

We are already installing like 4 git repos to install this project, why is this such a problem to link with a 5th one?

**Comment by rvalyi** [comment]
Instead I propose you something: if you guys manage to raise the pysped project code metrics to the ones of the OCA, then we will suport having the odoo-brazil-eletronic-documents entering in the OCA, but as a separate project. Not only this state clear interfaces as it's also much easier to manage. We will never bother you with the leadership of the odoo-brazil-eletronic-documents project as you already have a clear leadership on it, and we don't want to be bothered in the leadership in this project we did at 90% and have good clues how to do it.

Look at Github, the world is full of these independent repo with clear responsibilities and leadership. I think this is just easier to manage and let it evolve in modular way. Today pysped is what rocks. But may be in 2 years some Java o say Golang lib will do it better. If we bundle it altogether, we are kind of dismissing to find a more sustainable alternative (and I think it would be cool).

**Comment by rvalyi** [comment]
The pysped also depends on geraldo, yet an other unmaintained project...

Please explain me, what is the trouble with the project not being bundled together? If you miss something here, please tell us (forget the last 5 months), if we aren't reactive enough you branch it and get the interface you need to plug the pysped project. Not possible?

BTW, the issues you are complaining about in the l10n-brazil project are even more present in the pysped project, so why want to make you and us tied to it for the rest of our life?

**Comment by renatonlima** [comment]
Danimar,

I also contribute and help to maintain PySPED project
https://github.com/aricaldeira/PySPED/pulls?utf8=%E2%9C%93&q=is%3Apr+author%3Arenatonlima+
PySPED is
a fantastic project, but as you know, brasil fiscal requirements is very
complex and maybe this is part of problem to implement a stable and
trustable solution, in last 2 months I found 4 bugs in PySPED, I think
before we integrate this project in a OCA repo we must work to improve
PySPED to be a stable lib

2015-06-10 17:24 GMT-03:00 Danimar Ribeiro notifications@github.com:

**Comment by danimaribeiro** [comment]
I also agree  Renato that we have to improve a bit the pysped codebase, but that is not an impediment to do so.

dbms drivers and this are totally different things, you have different databases that needed different solutions, here we just one way to send via web service over http, with just one standard, so I don't know how that apply here.

"Please explain me, what is the trouble with the project not being bundled together? If you miss something here, please tell us (forget the last 5 months), if we aren't reactive enough you branch it and get the interface you need to plug the pysped project. Not possible?"

It's not about bundling together, is about to create a localization that is complete and have everything to run at least a small company. You talk about the purpose of OCA, but what's the purpose of localization? Just some modules to help use odoo in brazil but you have to implement a lot of things on your own?  I think we can't just improve code with tests in the localization, we have to create more customer value to it.

**Comment by danimaribeiro** [comment]
Just to be clear, it's not my intention to have this merge done tomorrow, but we can create a roadmap with the improvements, and put it in the backlog to work on. We have now kmee, akretion and probably in a few months I will start to work full time in Odoo, so what's the real impediment to improve this code and merge it together in v9 maybe?

**Comment by rvalyi** [comment]
@danimaribeiro, when I gave the Rails example I didn't express me well: I wanted to talk about the ActiveRecord adapters. ActiveRecord is central to Rails. But the ActiveRecords adapters are many, even for a single DB technology, yet depend on Rails and aren't bundled together.

With Odoo, there are people transmitting NF'e manually with the emissor, there are people doing txt importation in the free emissor, there are people who did and used at least for some times an open source transmissor for the Technospeed proprietary library, there are people using the Uninfe project. We want to leave that door open. So for instance if you want to use Odoo with Pypy and you bundle the pysped transmissor with this project you would not even be able to just install or test it with Pypy.

Yes the project with pypsed is by far the best one. But it doesn't mean it should be the only one.

Also, as I said since the beginning, I would prefer a client server architecture a bit like Uninfe than bundling everything inside the same obese runtime. A micro-service architecture would be far easier to maintain in the long run (like these possible future Python runtime incompatibilities). For instance that's the micro-service architecture that just saved the Aeroo reporting engine when Libreoffice got bundled with Python 3 when Odoo sticked to Python 2 for now.

And mostly I fail to see any benefit in having it into the same git repo. In the OCA we can have it in the future yes, but it has nothing to do inside the same repo as we want to preserve the project modularity.

Now about having it inside the OCA and having a roadmap for it. I would say on my side we need:
-  these project leaders (you or @mileo) should be known to the OCA in general. That is helping in this project, but gaining legitimacy in the Odoo ecosystem in general. Today I believe you never even reported or fixed an Odoo bug in the Odoo repo. I mean that kind of things, do OCA reviews and PR, so the OCA know you do solid work and you are here for serious reasons. We kind of trust it, but you should prove it to the OCA if you want to lead some OCA project.
-  The codebase of odoo-brazil-eletronic-documents should become close to the OCA standards: passing the standard Travis tests with some coverage, PEP-8, that's the easy part and Akretion can certainly help here.
-  We should clear up the project dependencies. Today for instance it depends on the pysped lib that is clearly not yet up to the OCA mark. May be improve the situation with pysped, ensure it has some test coverage, so bug tracker activity (hardly the case today, everybody fork it as they need, that's kind of a mess and I know we didn't do better here over the 6 last months but really that was exceptional). It's also a problem that it depends on Geraldo with no release since 2012 and a hard dependency on CPython 2.7 again I believe. Do you know that the OCA was rejecting project depending on Aeroo reports just because Aeroo wasn't part of the OCA? And sorry but the situation with Aeroo is in fact far more positive than with Geraldo. So I think there is no reason we should make an exception here in the OCA. Once the dependency picture looks more sustainable, then you have all my support to include it in the OCA, but as an extra project as I said.

That being said, if you cannot integrate properly with the l10n-brazil project because some interface is missing, please open a PR, propose an extension point and it's our job to provide it so you can plug properly on the project.

**Comment by danimaribeiro** [comment]
"With Odoo, there are people transmitting NF'e manually with the emissor, there are people doing txt importation in the free emissor, there are people who did and used at least for some times an open source transmissor for the Technospeed proprietary library, there are people using the Uninfe project."
They don't have a official way to do it, so they improvise.

I talked with some from OCA and as this project is specific to brazilian localization they don't find a good idea a new repo.

Odoo and Pypy is way too far yet.

"we want to preserve the project modularity." - You won't loose modularity with this project at least I don't see how.

"Today I believe you never even reported or fixed an Odoo bug in the Odoo repo. I mean that kind of things, do OCA reviews and PR, so the OCA know you do solid work and you are here for serious reasons. We kind of trust it, but you should prove it to the OCA if you want to lead some OCA project."
I did fix bugs in Odoo and OCA already, in OCA mostly in form of translations. But the localization has more work to be done than my free time, so I invest all my time in the localization, if I have more time I'll spend more in the OCA, but localization is also OCA.

**Comment by rvalyi** [comment]
I propose to have a meeting with the OCA board folks when it's question to merge the other project into this repo. I want this the technical and modularity aspect to be discussed by people who understand what we are talking about. Sorry but I still see it a bit as a hijack of the original project scope and I fail to see any technical benefit of doing so.

**Comment by rvalyi** [comment]
BTW, Odoo starts to work on Pypy (see in the community list), it's just slow at the moment. Python 3 migration will also become a question in the close future. By bundling pypsed together we will just make us a big drag on all that for no benefit IMHO.

**Comment by danimaribeiro** [comment]
I read all the thread about the Pypy.

2015-06-11 13:11 GMT-03:00 Raphaël Valyi notifications@github.com:

> BTW, Odoo starts to work on Pypy (see in the community list), it's just
> slow at the moment. Python 3 migration will also become a question in the
> close future. By bundling pypsed together we will just make us a big drag
> on all that for no benefit IMHO.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/OCA/l10n-brazil/issues/155#issuecomment-111188526.

## 

Danimar Ribeiro
Code https://github.com/danimaribeiro
Blog http://openerpbr.blogspot.com.br/

**Comment by rvalyi** [comment]
@danimaribeiro you claim you fixed Odoo bugs. I don't aim to be rude but I don't find you https://github.com/odoo/odoo/graphs/contributors (unlike Renato or me or other Akretion guys). As for translation we all do it. As for contributing to this project, we all do it. The point is that is seems you somewhat seems to question our project leadership. I say fair enough. I have been elected an OCA reviewer, we co-funded the OCA, you can very well find other OCA people to give you reason. But you'll probably convince them more easily by doing stuff also outside from this project so you don't demonstrate your skills and dedication just to us that was my point about legitimacy in the OCA.

Again, we are very aware we screw it up the last 5 months or so. Again we had NO CHOICE unfortunately, we would have broken to release earlier, we had no time to interact anymore and we couldn't afford relying on third party localization during the months of tension. As Renato will explain we heavily refactored the codebase Akretion had written at 95% and are bringing it up to the OCA standards (it's just marginally out today). The code was not that good before but at least it was always better than the Odoo core codebase, so once we did better than Odoo core, our effort was more on making the core better.

You said the localization is the priority. That's a point of view... I invite you to think what would be the situation of some people like us and others didn't create the OCA. Probably the open source localization would be pointless today in Brazil as in many countries...

In fact it's all a tradeoff. 6 to 4 years ago, we were only able to bootstrap the localization by working for Europe and the USA from Brazil. Including, Renato imported several fiscal concepts of the localization from designing the "intrastat" inter countries fiscal system in Europe for EU customers using OpenERP... 

Even today, there is on one side the requirements of the average Brazilian SMB like easy automatic transmission of NFe's. But on the other side there are the ways to achieve that goal. And often the ways to achieve it is to work for companies where pure transmission isn't the main point, so they can benefit from Odoo enough and pay us to actually work on improving the localization big picture... That"s why we don't focus on the pysped stuff even if we help and helped design the interface with this project...

**Comment by danimaribeiro** [comment]
I don't care who is the leadership or whatever.
I don't need to be OCA member to propose something new, or do I need to? If you guys don't want to accept ideas from people that don't participate in the OCA, i think you can close this issue and move on.

**Comment by rvalyi** [comment]
You don't need to be OCA or whatever to propose something new no. I think you will agree we ALWAYS considered all PR in this project except for the last 5 month or so because unfortunately we had a good reason for. Now things are normal again. If you catch us not taking care of third party ideas in the PR's, I encourage you to complain, to the OCA for instance.

Now, for this PR for instance where you would like to change the project scope and where we disagree. I would say that you would need to badly convince the OCA we are wrong about that so we change our mind. And a way to gain that support is by doing stuff also outside from the project.

But in any case I don't want to be rude, we may have a divergence here, we generally greatly appreciate your contribs in this project

**Comment by danimaribeiro** [comment]
I agree with you on this, but I want to convince you guys that this is necessary, I don't think we need to ask the other OCA members.

Here is my contribution:
https://github.com/odoo/odoo/issues/3879
But I also don't see Renato contributions there :)  just you, but that is because it show just 100 contributors.

**Comment by rvalyi** [comment]
OK. You are right Yeah Renato did hundreds of commits too, he doesn't appear in the top 100 either.

In any case I don't want to be rude, we may have a divergence here, we generally greatly appreciate your contribs in this project and will value them. For instance we included you explicitely as one of the main contributors in the project README. If you want us to put some link as well, please just tell us which link.

OK so you want to convince us it's necessary.
So let's develop: what exactly could you do if projects were merged that you cannot do today? Revision control? That's just about pinning git revisions in your buildout.cfg (we advise you use buildout).

**Comment by rvalyi** [comment]
(I meant Renato did dozens of commits only, that was a typo)

**Comment by mileo** [comment]
Hi,

Install PySPED is not so simple if you do not master the process, however install Odoo in production enviroment with custom and OCA requires a bit of knowledge.

For small businesses, who want to install Odoo for own account have PySPED and NFE together is a big hand on the wheel.

If this process is well defined is gonna easy too to large companies that have IT staff to install and test Odoo.

If the problem is Geraldo, we can make the report in Odoo. That even already done with Aeroo by Akretion, install Aeroo production i consider more problematic than PYSPED.

The PySPED will improve, is not perfect, but it's very handy.

Much has been validated and he can handle the stride. It needs some corrections, mainly in features that @aricaldeira does not use in production. For example the generation of a nfe object from a XML.

We are currently developing the module NFC-E and integration with SAT. It took several changes in nfe and including PySPED to make NFC-E work.

If the integrator wants not install nfe module and create your own module.

**But we really needed put an end to the issue of NF-E and left for other items in the SPED project.**

**Comment by danimaribeiro** [comment]
I would like to use all the OCA infrastructure to this module, runbot, the
PR would have reviews, there is documentation in the OCa on how to do
contributions, etc, the module will benefit from all of this.
Second, just one module of NF-e inside the core wouldn't break modularity,
if you don't want to install the nf-e automatic transmission you will be
able to export or do whatever you wants, like it is today.

2015-06-11 13:53 GMT-03:00 Raphaël Valyi notifications@github.com:

> OK. You are right Yeah Renato did hundreds of commits too, he doesn't
> appear in the top 100 either.
> 
> In any case I don't want to be rude, we may have a divergence here, we
> generally greatly appreciate your contribs in this project and will value
> them. For instance we included you explicitely as one of the main
> contributors in the project README. If you want us to put some link as
> well, please just tell us which link.
> 
> OK so you want to convince us it's necessary.
> So let's develop: what exactly could you do if projects were merged that
> you cannot do today? Revision control? That's just about pinning git
> revisions in your buildout.cfg (we advise you use buildout).
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/OCA/l10n-brazil/issues/155#issuecomment-111205368.

## 

Danimar Ribeiro
Code https://github.com/danimaribeiro
Blog http://openerpbr.blogspot.com.br/

**Comment by rvalyi** [comment]
So we can have Travis already, I can make a PR for that.
As for Runbot we wouldn't have it yet indeed, or well that should be investigated if it's possible or not.

But IMHO what we need is indeed to bring the project into the OCA (provided it can clear a few doubts), but as a different repo related to the localization, but not the same.

Today I'm more concerned by people active in the BR community not being active elsewhere in the OCA than trying to expand the scope of the localization inside the OCA. Otherwise OCA doesn't mean anything, it would mean bring any project we want and that would not be homogeneous, it would mean nothing except very hard to manage. It's the size vs quality question, but sorry today I feel the BR community is really isolated aside from us. I see European communities way more integrated to the OCA activity for instance. Hopefully this changes. Certainly the reason is because the localization issue was so hard in Brazil that few specialized companies could make a living working with it.

So I'll weight for the project inclusion in the OCA, as an additional repo (I'm sure we can have a talk about that with other OCA guys), provided the dependency picture clears up a bit.

Look, we grouped modules in the OCA, because 4 years ago, all community modules who wanted to catch attention were grouped inside e single repo and because the Odoo addons themselves have this monolithic design.

But it's rather a big design flaw! and we are aware of that at the OCA. The trend is more that we will move each module into a different git project in the future than group them even more than today.

The reason why it's like that today is IMHO more or less:
-  a bad package management in Python in general, say compared to Ruby or some other languages (not as bad as Java though ;-)
-  an ignorance from Odoo SA how to package stuff properly the Python way. Now they do have some good Python programmers, but it has not always be the case. 4 years ago they had mostly no clue at all how to do that properly.
-  a deliberate willingness to plan to promote a central editor controlled appstore rather than having people install doing just pip install

Then as we don't have a proper way (Buildout helps but that's still weak) to package and install modules in a given revision if they belong to different repos, we keep grouping them artificially the way we do. But the trend is toward packaging properly and having different repos.

If you look for instance the Rails ecosystem for instance, it's really common to have a project installing around 100 dependencies and they all live into a different git repo and it's all pinned, included for dev and it just work and people just do that with many other languages/framework, it's the pattern that work for open source because it fosters modularity, both in the code and in the management of it.

But the trend today is toward separation so I hope we don't go counter logic here.

**Comment by danimaribeiro** [comment]
You know that the localization is the entrance of OCA for the programmers here in Brazil, don't expect someone to contribute to OCA before he is able to contribute to localization.
I acknowledge your concerns about the OCA and everything, would be better to have more contributions at all, but they need to start somewhere.

According to create a new repo to the nf-e in the OCA, I don't mind. As I said would be nice to have this project under OCA, I really like the idea of OCA to maintain and protect the odoo modules created by the community. 

A lot of people are using the nf-e module, and if we transfer to OCA, OCA also will have more visibility in Brasil, maybe will help what you are aiming for.

**Comment by rvalyi** [comment]
@danimaribeiro I think this a bad idea for Brazilian to start trying to work with Odoo in Brazil. The 1st years are though. Today if you take the US market for instance, it just lack Odoo developers the same way as in Brazil (a bit less but demand is higher too). It's just an advice, but it's way easier today to be paid as an Odoo developer working for the USA from Brazil than trying to get a Brazilian SMB pay you decent money for that. It doesn't mean it's bad to try working for Brazil (today we do 100% here), but it's just harder. But at the beginning of Akretion, we worked 1 year 100% for outside and then may be 80% for outside. Even today, if we were just thinking to money we would earn 3x more working remote for outside.

Also I think many make a big mistake: they believe Odoo marketing it targets traditional out of the box management software (because it's the interest of Odoo SA to market it this way). But it's not true, nearly all people in the OCA do "framework jobs" instead to pay the bill: that is customize Odoo a lot to meet specific requirements. Here the fiscal aspect of Odoo can also be a bit less of a problem in Brazil.

Finally, I also feared that if we abolish the natural boundaries of the project repo, we will start to see contribs with the l10n_br_account module trying to use pypsed directly and this is were I fear that would taint the whole project. If somebody need to do that, today they need to ask us (or propose) new hooks so that they can plug thei Pysped stuff. But that would be dangerous not having that architecture IMHO.
